File: backend/app/schemas/games_admin.py
"""Live game facilitator-authoring schemas (Live Games Phase 2C, 8-3)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class GameQuestionOptionIn(BaseModel):
    text: str
    is_correct: bool = False


class GameQuestionIn(BaseModel):
    prompt: str
    options: list[GameQuestionOptionIn] = Field(min_length=2)
    time_limit_seconds: int | None = None  # None = use the game's default
    points_mode: Literal["normal", "double"] = "normal"
    position: int | None = None  # None = append at the end

    @model_validator(mode="after")
    def _validate_correct_option(self) -> "GameQuestionIn":
        correct = sum(1 for o in self.options if o.is_correct)
        if correct != 1:
            raise ValueError(f"a game question needs exactly one correct option, got {correct}")
        return self


class GameQuestionUpdate(BaseModel):
    prompt: str | None = None
    options: list[GameQuestionOptionIn] | None = None
    time_limit_seconds: int | None = None
    points_mode: Literal["normal", "double"] | None = None
    position: int | None = None

    @model_validator(mode="after")
    def _validate_correct_option(self) -> "GameQuestionUpdate":
        if self.options is not None:
            correct = sum(1 for o in self.options if o.is_correct)
            if correct != 1:
                raise ValueError(f"a game question needs exactly one correct option, got {correct}")
        return self


class GameQuestionOut(BaseModel):
    id: UUID
    position: int
    prompt: str
    options: list[GameQuestionOptionIn]
    time_limit_seconds: int | None
    points_mode: str
    max_points: int  # resolved via services/games/points.py — the frontend never computes this itself


class GameQuestionReorderIn(BaseModel):
    question_ids: list[UUID]


class GameCreate(BaseModel):
    title: str
    description: str | None = None
    default_time_limit_seconds: int = Field(gt=0, default=20)
    default_floor_pct: int = Field(ge=0, le=100, default=25)
    default_blackout_count: int = Field(ge=0, default=3)


class GameUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    default_time_limit_seconds: int | None = Field(gt=0, default=None)
    default_floor_pct: int | None = Field(ge=0, le=100, default=None)
    default_blackout_count: int | None = Field(ge=0, default=None)


class GameOut(BaseModel):
    id: UUID
    title: str
    description: str | None
    created_by: UUID
    default_time_limit_seconds: int
    default_floor_pct: int
    default_blackout_count: int
    created_at: datetime | None
    question_count: int


class GameDetailOut(GameOut):
    questions: list[GameQuestionOut]

File: frontend/src/api/games_admin.ts
/** Live game facilitator-authoring API (Live Games Phase 2C, 8-3) — thin
 * wrapper over `/games/admin/*`. Types mirror `schemas/games_admin.py`. */
import { api } from "@/api/client";

export type PointsMode = "normal" | "double";

export interface GameQuestionOption {
  text: string;
  is_correct: boolean;
}

export interface GameQuestion {
  id: string;
  position: number;
  prompt: string;
  options: GameQuestionOption[];
  time_limit_seconds: number | null;
  points_mode: PointsMode;
  max_points: number;
}

export interface Game {
  id: string;
  title: string;
  description: string | null;
  created_by: string;
  default_time_limit_seconds: number;
  default_floor_pct: number;
  default_blackout_count: number;
  created_at: string | null;
  question_count: number;
}

export interface GameDetail extends Game {
  questions: GameQuestion[];
}

export interface GameCreateInput {
  title: string;
  description?: string | null;
  default_time_limit_seconds?: number;
  default_floor_pct?: number;
  default_blackout_count?: number;
}

export type GameUpdateInput = Partial<GameCreateInput>;

export interface GameQuestionInput {
  prompt: string;
  options: GameQuestionOption[];
  time_limit_seconds?: number | null;
  points_mode?: PointsMode;
  position?: number;
}

export type GameQuestionUpdateInput = Partial<GameQuestionInput>;

export const listGamesApi = () => api.get<Game[]>("/games/admin").then((r) => r.data);

export const createGameApi = (data: GameCreateInput) =>
  api.post<Game>("/games/admin", data).then((r) => r.data);

export const getGameApi = (gameId: string) =>
  api.get<GameDetail>(`/games/admin/${gameId}`).then((r) => r.data);

export const updateGameApi = (gameId: string, data: GameUpdateInput) =>
  api.patch<Game>(`/games/admin/${gameId}`, data).then((r) => r.data);

export const deleteGameApi = (gameId: string) =>
  api.delete<void>(`/games/admin/${gameId}`).then((r) => r.data);

export const createGameQuestionApi = (gameId: string, data: GameQuestionInput) =>
  api.post<GameQuestion>(`/games/admin/${gameId}/questions`, data).then((r) => r.data);

export const updateGameQuestionApi = (questionId: string, data: GameQuestionUpdateInput) =>
  api.patch<GameQuestion>(`/games/admin/questions/${questionId}`, data).then((r) => r.data);

export const duplicateGameQuestionApi = (questionId: string) =>
  api.post<GameQuestion>(`/games/admin/questions/${questionId}/duplicate`).then((r) => r.data);

export const deleteGameQuestionApi = (questionId: string) =>
  api.delete<void>(`/games/admin/questions/${questionId}`).then((r) => r.data);

export const reorderGameQuestionsApi = (gameId: string, questionIds: string[]) =>
  api.post<GameQuestion[]>(`/games/admin/${gameId}/questions/reorder`, { question_ids: questionIds }).then((r) => r.data);
